tf2onnx/optimizer/__init__.py

```diff
@@ -6,43 +6,51 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
-import sys
-import traceback
 from collections import OrderedDict
+import copy
 
-from tf2onnx.optimizer.const_fold_optimizer import ConstFoldOptimizer
-from tf2onnx.optimizer.identity_optimizer import IdentityOptimizer
-from tf2onnx.optimizer.merge_duplicated_nodes_optimizer import MergeDuplicatedNodesOptimizer
-from tf2onnx.optimizer.transpose_optimizer import TransposeOptimizer
-
-# pylint: disable=missing-docstring, broad-except
+from .const_fold_optimizer import ConstFoldOptimizer
+from .identity_optimizer import IdentityOptimizer
+from .merge_duplicated_nodes_optimizer import MergeDuplicatedNodesOptimizer
+from .transpose_optimizer import TransposeOptimizer
+from .. import logging
 
 # optimizer sequence need to be considered carefully
 _optimizers = OrderedDict([
-    ("transpose_opt", TransposeOptimizer),
-    ("fold_const", ConstFoldOptimizer),
-    # merge_duplicated_nodes should be used after transpose_opt
-    # for transpose_opt may have some trans nodes that can be merge
-    ("merge_duplicated_nodes", MergeDuplicatedNodesOptimizer),
-    ("identity_opt", IdentityOptimizer),
+    ("reduce_transpose", TransposeOptimizer),
+    ("fold_constants", ConstFoldOptimizer),
+    # merge_duplication should be used after reduce_transpose
+    # for reduce_transpose may have some trans nodes that can be merge
+    ("merge_duplication", MergeDuplicatedNodesOptimizer),
+    ("reduce_identity", IdentityOptimizer),
 ])
 
 
-def optimize_graph(graph):
-    try:
-        opts = _get_optimizers()
-        for opt in opts.values():
-            graph = opt().optimize(graph)
-
-        graph.update_proto()
-        return graph
-    except Exception:
-        # degradation to non-optimized model proto
-        type_, value_, traceback_ = sys.exc_info()
-        ex_ext = traceback.format_exception(type_, value_, traceback_)
-        print("NON-CRITICAL error in optimizer: ", ex_ext)
-        return None
-
-
 def _get_optimizers():
     return _optimizers
+
+
+def optimize_graph(graph):
+    """ Optimize graph, return optimized graph. No throw. """
+    logger = logging.getLogger(__name__)
+    logger.info("Optimizing ONNX model")
+
+    before = graph.dump_node_statistics()
+    opts = _get_optimizers()
+    for name, factory in opts.items():
+        try:
+            logger.verbose("Apply %s", name)
+            current = copy.deepcopy(graph)
```

Comment:
this graph can be copied completely including nodes/shapes/contained subgraphs? have your verified that?

Reply:
seems so.

https://docs.python.org/3/library/copy.html#copy.deepcopy

```
The difference between shallow and deep copying is only relevant for compound objects (objects that contain other objects, like lists or class instances):

A shallow copy constructs a new compound object and then (to the extent possible) inserts references into it to the objects found in the original.
A deep copy constructs a new compound object and then, recursively, inserts copies into it of the objects found in the original.
```